cmd/lambda/usage/list.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Optum/dce/pkg/api/response"
+	"github.com/Optum/dce/pkg/usage"
+)
+
+// GetUsage - Gets all of the usage
+func GetUsage(w http.ResponseWriter, r *http.Request) {
+
+	// This has become a "fall-through" method for any of the URL combinations that
+	// don't match the explicit routes, so we parse input here to get all of the
+	// query string values that are supplied on the URL
+	getUsageInput, err := parseGetUsageInput(r)
+
+	if err != nil {
+		response.WriteRequestValidationError(w, fmt.Sprintf("Error parsing query params"))
+		return
+	}
+
+	result, err := UsageSvc.GetUsage(getUsageInput)
+
+	if err != nil {
+		response.WriteServerErrorWithResponse(w, fmt.Sprintf("Error querying usage: %s", err))
+		return
+	}
+
+	// Serialize them for the JSON response.
+	usageResponseItems := []response.UsageResponse{}
+	for _, usageItem := range result.Results {
+		usageResponseItems = append(usageResponseItems, response.UsageResponse(*usageItem))
+	}
+
+	// If the DB result has next keys, then the URL to retrieve the next page is put into the Link header.
+	if len(result.NextKeys) > 0 {
+		nextURL := buildNextURL(r, result.NextKeys)
+		w.Header().Add("Link", fmt.Sprintf("<%s>; rel=\"next\"", nextURL))
+	}
+
+	json.NewEncoder(w).Encode(usageResponseItems)
+}
+
+// parseGetUsageInput creates a GetUsageInput from the query parameters
+func parseGetUsageInput(r *http.Request) (usage.GetUsageInput, error) {
+	query := usage.GetUsageInput{
```

Comment:
We have a number of these "parse..Input()" methods. I'm wondering if we can make these a bit more dynamic and use reflection? It looks like this is already part of the gorilla toolkit (see https://stackoverflow.com/questions/40564842/convert-url-query-map-of-slices-to-struct-golang).

Reply:
I'm going to be submitting a pull request very shortly that does this.

Reply:
Thanks

Reply:
See https://github.com/kddejong/dce/blob/Feature/AccountControllerRedux/pkg/model/account.go#L4-L13
and
https://github.com/kddejong/dce/blob/Feature/AccountControllerRedux/cmd/lambda/accounts/list.go#L45
then
https://github.com/kddejong/dce/blob/8c9ff115c26ebe3938d1fd94902da0e5d88b3e1a/pkg/api/helpers.go#L12